node/node.go

```diff
@@ -0,0 +1,452 @@
+	workerKey := blake2b.Sum256(append([]byte("worker"), walletKey...))
+	w := worker.New(workerKey, b)
+	return w, func() error { return nil }, nil
+}
+
+func newAutopilot(cfg AutopilotConfig, dir string) (*autopilot.Autopilot, func() error, error) {
+	store, err := stores.NewJSONAutopilotStore(dir)
+	if err != nil {
+		return nil, nil, err
+	}
+	b := bus.NewClient(cfg.BusAddr, cfg.BusPassword)
+	w := worker.NewClient(cfg.WorkerAddr, cfg.WorkerPassword)
+	a, err := autopilot.New(store, b, w, cfg.LoopInterval)
+	if err != nil {
+		return nil, nil, err
+	}
+	cleanup := func() error {
+		a.Stop()
+		return nil
+	}
+	return a, cleanup, nil
+}
+
+// Node describes a single renterd process with one or more enabled components.
+// A node always needs to either create a bus or connect to an existing one by
+// calling CreateBus or AddBus respectively.
+// Afterwards a worker can be created or added with CreateWorker/AddWorker. If
+// the worker is created, it will automatically connect to the known bus. If it
+// is added, the caller needs to guarantee that the worker is connected to the
+// same bus as specified in the node.
+// After both a bus and a worker were added, the autopilot can be created. It
+// will connect to the bus and worker from the previous steps.
+type Node struct {
+	busAddr     *string
+	busPassword *string
+	bus         *bus.Bus
+
+	workerAddr     *string
+	workerPassword *string
+	w              *worker.Worker
+
+	autopilotAddr     *string
+	autopilotPassword *string
+	a                 *autopilot.Autopilot
+
+	apiAddr     string
+	apiPassword string
+	auth        func(http.Handler) http.Handler
+	dir         string
+	l           net.Listener
+	wk          consensus.PrivateKey
+
+	mux TreeMux
+	srv http.Server
+
+	cleanupFuncs []func() error
+}
+
+// NewNode creates a new, empty node without any components. The components are
+// then added through the various exposed methods. For a more detailed
+// description take a look at the Node's type documentation.
+func NewNode(apiAddr, apiPassword, dir string, uiHandler http.Handler, wk consensus.PrivateKey) *Node {
+	// Start listening.
+	l, err := net.Listen("tcp", apiAddr)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer l.Close()
+
+	mux := TreeMux{
+		H:   uiHandler,
+		Sub: make(map[string]http.Handler),
+	}
+
+	// Overwrite APIAddr now that we know the exact addr from the listener.
+	return &Node{
+		apiAddr:     "http://" + l.Addr().String(),
+		apiPassword: apiPassword,
+		auth:        jape.BasicAuth(apiPassword),
+		dir:         dir,
+		l:           l,
+		mux:         mux,
+		srv:         http.Server{Handler: mux},
+		wk:          wk,
+	}
+}
+
+// APIAddress returns the node's API address.
+func (c *Node) APIAddress() string {
+	return c.apiAddr
+}
+
+// Close shuts down the API as well as the components that were created.
+// Remote components that were added instead of created will remain active.
+func (c *Node) Close() error {
```

Comment:
`Close` could take a context and all cleanup functions could pass it and return an error so we can properly apply a timeout and log the error if it failed to shut down gracefully

Reply:
I'm against changing `Close` to not match `io.Closer` anymore. But I'm open to renaming it to `Shutdown(context)` to get the context in there.